File: Simulator/DMSimulator/DMTasks/ConfigurationUpdate.c
#include "../SendReport.h"
#include "../Utilities.h"
#include "ConfigurationUpdate.h"

// State switch graph: pending -> downloading -> applying -> idle
static DMTaskStep _steps[] =
{
	{ CU_PENDING, 0, CU_DOWNLOADING },
	{ CU_DOWNLOADING, 10, CU_APPLYING },
	{ CU_APPLYING, 10, DM_IDLE },
	{ DM_NULL, 0, DM_NULL }
};

static char* _version;
static time_t _downloadStartTime = 0;
static time_t _applyStartTime = 0;

static BOOL OnEnterState(DMTaskState state, time_t now)
{
	BOOL succeeded = TRUE;

	unsigned char* report = NULL;
	size_t size = 0;

	switch (state)
	{
	case CU_PENDING:
		// No report for entering pending state
		break;

	case CU_DOWNLOADING:
		_downloadStartTime = now;

		if (strcmp(_version, "downloadFail") == 0)
		{
			succeeded = FALSE;

			AllocAndPrintf(
				&report,
				&size,
				"{ 'iothubDM': { 'configurationUpdate': { 'status': 'download failed', 'log': 'download failed' } } }");
		}
		else
		{
			AllocAndPrintf(
				&report,
				&size,
				"{ 'iothubDM': { 'configurationUpdate': { 'status': 'downloading', 'log': 'downloading' } } }");
		}
		break;

	case CU_APPLYING:
		_applyStartTime = 0;

		if (strcmp(_version, "applyFail") == 0)
		{
			succeeded = FALSE;

			AllocAndPrintf(
				&report,
				&size,
				"{ 'iothubDM': { 'configurationUpdate': { 'status': 'apply failed', 'log': 'downloaded(%I64ds) -> apply failed' } } }",
				now - _downloadStartTime);
		}
		else
		{
			AllocAndPrintf(
				&report,
				&size,
				"{ 'iothubDM': { 'configurationUpdate': { 'status': 'applying', 'log': 'downloaded(%I64ds) -> applying' } } }",
				now - _downloadStartTime);
		}
		break;

	case DM_IDLE:
		AllocAndPrintf(
			&report,
			&size,
			"{ 'iothubDM': { 'configurationUpdate': { 'status': 'updated to %s', 'log': 'downloaded(%I64ds) -> applied(%I64ds)' } }, 'ConfigurationVersion': '%s' }",
			_version,
			_applyStartTime - _downloadStartTime,
			now - _applyStartTime,
			_version);
		break;

	default:
		printf("Unknown ConfigurationUpdateState %d", state);
		break;
	}

	if (report != NULL)
	{
		SendReport(report, size);
	}

	return succeeded;
}

static void OnLeaveState(DMTaskState state, time_t now)
{
	unsigned char* report = NULL;
	size_t size = 0;

	switch (state)
	{
	case DM_IDLE:
		// No report for leaving idle state
		break;

	case CU_PENDING:
		// No report for leaving pending state
		break;

	case CU_DOWNLOADING:
		AllocAndPrintf(
			&report,
			&size,
			"{ 'iothubDM': { 'configurationUpdate': { 'status': 'download completed', 'log': 'downloaded(%I64ds)' } } }",
			now - _downloadStartTime);
		break;

	case CU_APPLYING:
		AllocAndPrintf(
			&report,
			&size,
			"{ 'iothubDM': { 'configurationUpdate': { 'status': 'apply completed', 'log': 'downloaded(%I64ds) -> applied(%I64ds)' } } }",
			_applyStartTime - _downloadStartTime,
			now - _applyStartTime);
		break;

	default:
		printf("Unknown ConfigurationUpdateState %d", state);
		break;
	}

	if (report != NULL)
	{
		SendReport(report, size);
	}
}

BOOL BeginConfigurationUpdate(const char* uri, DMTaskStep** steps, OnEnterStateProc* onEnterStateProc, OnLeaveStateProc* onLeaveStateProc)
{
	// [WORKAROUND] Directly pick the version from the URI
	const char* prefix = "/configuration/";
	char* p = strstr(uri, prefix);
	if (p == NULL)
	{
		return FALSE;
	}
	p += strlen(prefix);

	free(_version);
	_version = _strdup(p);

	*steps = _steps;
	*onEnterStateProc = OnEnterState;
	*onLeaveStateProc = OnLeaveState;
	return TRUE;
}

File: Simulator/Simulator.WebJob/SimulatorCore/Devices/DMTasks/ConfigurationUpdate.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using Microsoft.Azure.Devices.Applications.RemoteMonitoring.Common.Extensions;
using Microsoft.Azure.Devices.Applications.RemoteMonitoring.Simulator.WebJob.SimulatorCore.Transport;
using Microsoft.Azure.Devices.Client;
using Microsoft.Azure.Devices.Shared;
using Newtonsoft.Json;

namespace Microsoft.Azure.Devices.Applications.RemoteMonitoring.Simulator.WebJob.SimulatorCore.Devices.DMTasks
{
    class ConfigurationUpdate : DMTaskBase
    {
        static private string LogPath = "Method.ConfigurationUpdate.Log";

        private Stopwatch _watch;
        private LogBuilder _logBuilder = new LogBuilder();

        public string Uri { get; private set; }
        public string Version { get; private set; }

        public ConfigurationUpdate(MethodRequest request)
        {
            var payload = JsonConvert.DeserializeObject<dynamic>(request.DataAsJson);

            var uri = (string)payload.ConfigUri;
            if (string.IsNullOrWhiteSpace(uri))
            {
                throw new ArgumentException("Missing ConfigUri");
            }

            // [WORKAROUND] Directly pick the version from the URI
            var match = new Regex("/configuration/(?<version>.*)$").Match(uri);
            if (!match.Success)
            {
                throw new ArgumentException("Bad format of ConfigUri");
            }

            Uri = uri;
            Version = match.Groups["version"].Value;

            // State switch graph: pending -> downloading -> applying -> idle
            _steps = new List<DMTaskStep>
            {
                new DMTaskStep { CurrentState = DMTaskState.CU_PENDING, ExecuteTime = TimeSpan.Zero, NextState = DMTaskState.CU_DOWNLOADING },
                new DMTaskStep { CurrentState = DMTaskState.CU_DOWNLOADING, ExecuteTime = TimeSpan.FromSeconds(10), NextState = DMTaskState.CU_APPLYING },
                new DMTaskStep { CurrentState = DMTaskState.CU_APPLYING, ExecuteTime = TimeSpan.FromSeconds(10), NextState = DMTaskState.DM_IDLE }
            };
        }

        protected override async Task<bool> OnEnterStateProc(DMTaskState state, ITransport transport)
        {
            bool succeed = true;
            var report = new TwinCollection();
            string status;

            switch (state)
            {
                case DMTaskState.CU_PENDING:
                    report = null;  // No report for entering pending state
                    break;

                case DMTaskState.CU_DOWNLOADING:
                    _watch = Stopwatch.StartNew();
                    succeed = Version != "downloadFail";
                    status = succeed ? "Downloading" : "Dowload failed";
                    report.Set(LogPath, _logBuilder.Append(status, succeed));
                    break;

                case DMTaskState.CU_APPLYING:
                    _watch = Stopwatch.StartNew();
                    succeed = Version != "applyFail";
                    status = succeed ? "Applying" : "Apply failed";
                    report.Set(LogPath, _logBuilder.Append(status, succeed));
                    break;

                case DMTaskState.DM_IDLE:
                    report.Set(DeviceBase.ConfigurationVersionPropertyName, Version);
                    break;

                default:
                    throw new ArgumentOutOfRangeException(nameof(state));
            }

            if (report != null)
            {
                await transport.UpdateReportedPropertiesAsync(report);
            }

            return succeed;
        }

        protected override async Task<bool> OnLeaveStateProc(DMTaskState state, ITransport transport)
        {
            var report = new TwinCollection();

            switch (state)
            {
                case DMTaskState.DM_IDLE:
                    report = null;  // No report for leaving idle state
                    break;

                case DMTaskState.CU_PENDING:
                    report = null;  // No report for leaving pending state
                    break;

                case DMTaskState.CU_DOWNLOADING:
                    report.Set(LogPath, _logBuilder.Append($"Downloaded({(int)_watch.Elapsed.TotalSeconds}s)"));
                    break;

                case DMTaskState.CU_APPLYING:
                    report.Set(LogPath, _logBuilder.Append($"Applied({(int)_watch.Elapsed.TotalSeconds}s)"));
                    break;

                default:
                    throw new ArgumentOutOfRangeException(nameof(state));
            }

            if (report != null)
            {
                await transport.UpdateReportedPropertiesAsync(report);
            }

            return true;
        }
    }
}